anchore_engine/services/catalog/catalog_impl.py

```diff
@@ -2196,32 +2191,85 @@ def _delete_image_for_real(userId, image_record, dbsession, image_ids, image_ful
+            try:
+                logger.debug("DELETEing image from archive %s/%s", bucket, archive_id)
+                rc = obj_store.delete(account_id, bucket, archive_id)
+            except Exception:
+                logger.exception(
+                    "Error deleting image from archive %s/%s", bucket, archive_id
+                )
 
-    logger.debug("DELETEing image from policy_engine")
 
+def _delete_image_catalog(account_id, image_digest, image_ids, full_tags, db_session):
+    try:
+        logger.debug("DELETEing image from catalog")
+        rc = db_catalog_image.delete(image_digest, account_id, session=db_session)
+    except Exception:
+        logger.exception("Error deleting image digest %s from catalog", image_digest)
+
+
+def _delete_image_policy_engine(
+    account_id, image_digest, image_ids, full_tags, db_session
+):
     # prepare inputs
+
     try:
-        pe_client = internal_client_for(PolicyEngineClient, userId=userId)
+        pe_client = internal_client_for(PolicyEngineClient, userId=account_id)
         for img_id in set(image_ids):
             logger.debug(
-                "DELETING image from policy engine userId = {} imageId = {}".format(
-                    userId, img_id
-                )
+                "DELETING image from policy engine account_id=%s image_id=%s",
+                account_id,
+                img_id,
             )
-            rc = pe_client.delete_image(user_id=userId, image_id=img_id)
-    except:
-        logger.exception("Failed deleting image from policy engine")
-        raise
+            rc = pe_client.delete_image(user_id=account_id, image_id=img_id)
+    except Exception:
+        logger.exception(
+            "Failed deleting image digest %s from policy engine", image_digest
+        )
+
+
+# collection of description and delete function tuples. Each delete that take account_id, image_digest, image_ids, full_tags and db_session as arguments
+# TODO hack to support extensions, come up with a better (oo) solution to override and extend behaviour
+image_gc_functions = [
+    ("artifacts", _delete_image_artifacts),
+    ("catalog record", _delete_image_catalog),
+    ("policy-engine record", _delete_image_policy_engine),
+]
+
+
+def _delete_image_for_real(userId, image_record, dbsession, image_ids, image_fulltags):
+    image_digest = image_record["imageDigest"]
+    logger.debug(
+        "Begin image deletion for account_id=%s, digest=%s, tags=%s, ids=%s",
+        userId,
+        image_digest,
+        image_fulltags,
+        image_ids,
+    )
+    for desc, delete_func in image_gc_functions:
+        try:
+            logger.debug("Executing delete for image %s", desc)
+            delete_func(
+                account_id=userId,
+                image_digest=image_digest,
+                image_ids=image_ids,
+                full_tags=image_fulltags,
+                db_session=dbsession,
+            )
+        except Exception:
```

Comment:
Exception is too broad this should be ValueError or extending the exceptions.py module to include a new Type that is more specific to this kind of error.